File: src/main/java/user11681/limitless/asm/LimitlessTransformer.java
package user11681.limitless.asm;

import java.util.ListIterator;
import net.gudenau.lib.unsafe.Unsafe;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.InsnList;
import org.objectweb.asm.tree.InsnNode;
import org.objectweb.asm.tree.IntInsnNode;
import org.objectweb.asm.tree.MethodInsnNode;
import org.objectweb.asm.tree.MethodNode;
import org.objectweb.asm.tree.TypeInsnNode;
import org.objectweb.asm.tree.VarInsnNode;
import user11681.fabricasmtools.plugin.TransformerPlugin;
import user11681.limitless.enchantment.EnchantingBlocks;
import user11681.limitless.enchantment.EnchantmentUtil;
import user11681.limitless.enchantment.EnchantmentWrapper;
import user11681.reflect.Accessor;
import user11681.shortcode.Shortcode;
import user11681.shortcode.instruction.ExtendedInsnList;

public class LimitlessTransformer extends TransformerPlugin implements Opcodes {
    private static void transformEnchantmentHelperGenerateEnchantments(MethodNode method) {
        ListIterator<AbstractInsnNode> iterator = method.instructions.iterator();

        Shortcode.findForward(iterator,
            (AbstractInsnNode instruction) -> instruction.getType() == AbstractInsnNode.INT_INSN && ((IntInsnNode) instruction).operand == 50,
            () -> {
                iterator.remove();

                iterator.add(new VarInsnNode(ILOAD, 2));
                iterator.add(new InsnNode(ICONST_2));
                iterator.add(new InsnNode(IDIV));
                iterator.add(new IntInsnNode(BIPUSH, 20));
                iterator.add(new InsnNode(IADD));
            }
        );
    }

    @Override
    public void onLoad(String mixinPackage) {
        super.onLoad(mixinPackage);

        this.putClass("World", 1937);
        this.putClass("BlockPos", 2338);
        this.putInternal("Enchantment", 1887);
        this.putField("creativeMode", 7477);
        this.putMethod("calculateRequiredExperienceLevel", 8227);
        this.putMethod("getMaxLevel", 8183);

        this.registerPostMixinMethodTransformer(klass(1890), method(8230), null, LimitlessTransformer::transformEnchantmentHelperGenerateEnchantments);
        this.registerPostMixinMethodTransformer(klass(3853, 1648), method(7246), null, this::transformEnchantBookFactoryCreate);
        this.registerPostMixinMethodTransformer(klass(1890), method(8229), null, this::transformEnchantmentHelperGetPossibleEntries);
        this.registerPostMixinMethodTransformer(klass(1718), method(17411), null, this::transformEnchantmentScreenHandler);
    }

    private void transformEnchantBookFactoryCreate(MethodNode method) {
        Shortcode.findForward(method.instructions.iterator(),
            (AbstractInsnNode instruction) -> instruction.getOpcode() == INVOKEVIRTUAL && ((MethodInsnNode) instruction).name.equals(this.method("getMaxLevel")),
            (MethodInsnNode instruction) -> {
                instruction.owner = EnchantmentWrapper.INTERNAL_NAME;
                instruction.name = "originalMaxLevel";

                method.instructions.insertBefore(instruction, new TypeInsnNode(Opcodes.CHECKCAST, EnchantmentWrapper.INTERNAL_NAME));
            }
        );
    }

    private void transformEnchantmentHelperGetPossibleEntries(MethodNode method) {
        InsnList instructions = method.instructions;
        ListIterator<AbstractInsnNode> iterator = instructions.iterator();

        Shortcode.findForward(iterator,
            (AbstractInsnNode instruction) -> instruction.getOpcode() == INVOKEVIRTUAL && ((MethodInsnNode) instruction).name.equals(this.method("getMaxLevel")),
            (MethodInsnNode instruction) -> {
                Shortcode.removeBetweenInclusive(iterator, AbstractInsnNode.LINE, AbstractInsnNode.IINC_INSN);

                iterator.next();
                iterator.remove();

                instructions.insert(iterator.previous(), new ExtendedInsnList()
                    .iload(0)
                    .aload(7)
                    .aload(3)
                    .invokestatic(
                        EnchantmentUtil.INTERNAL_NAME,
                        "getHighestSuitableLevel",
                        Shortcode.composeMethodDescriptor("V", "I", this.internal("Enchantment"), "java/util/List"),
                        true
                    )
                );
            }
        );
    }

    private void transformEnchantmentScreenHandler(MethodNode method) {
        ListIterator<AbstractInsnNode> iterator = method.instructions.iterator();

        Shortcode.findForward(iterator,
            (AbstractInsnNode instruction) -> instruction.getOpcode() == ICONST_0,
            (InsnNode instruction) -> {
                ((VarInsnNode) instruction.getNext()).setOpcode(FSTORE);
                iterator.set(new MethodInsnNode(
                    INVOKESTATIC,
                    EnchantingBlocks.INTERNAL_NAME,
                    "countEnchantingPower",
                    Shortcode.composeMethodDescriptor("F", this.klass("World"), this.klass("BlockPos")),
                    true
                ));
                iterator.previous();
                iterator.add(new VarInsnNode(ALOAD, 2));
                iterator.add(new VarInsnNode(ALOAD, 3));
            }
        );

        iterator.next();
        iterator.next();

        int gotoCount = 0;

        while (gotoCount != 3) {
            final AbstractInsnNode next = iterator.next();

            if (next.getOpcode() == GOTO) {
                ++gotoCount;
            }

            iterator.remove();
        }

        Shortcode.findForward(iterator,
            (AbstractInsnNode instruction) -> instruction.getOpcode() == ILOAD && ((VarInsnNode) instruction).var == 4,
            (VarInsnNode instruction) -> instruction.setOpcode(FLOAD)
        );

        Shortcode.findForward(iterator,
            (AbstractInsnNode instruction) -> instruction.getOpcode() == INVOKESTATIC,
            (MethodInsnNode instruction) -> {
                instruction.owner = EnchantingBlocks.INTERNAL_NAME;
                instruction.name = "calculateRequiredExperienceLevel";
                instruction.desc = instruction.desc.replaceFirst("II", "IF");
                instruction.itf = true;
            }
        );
    }

    static {
        Object delegate = Accessor.getObject(LimitlessTransformer.class.getClassLoader(), "delegate");
        LimitlessFabricMixinTransformerProxy transformer = Unsafe.allocateInstance(LimitlessFabricMixinTransformerProxy.class);

        transformer.original = Accessor.getObject(delegate, "mixinTransformer");

        Accessor.putObject(delegate, "mixinTransformer", transformer);
    }
}

File: src/net/auoeke/limitless/transform/LimitlessTransformer.kt
package net.auoeke.limitless.transform

import net.auoeke.extensions.asm.*
import net.auoeke.extensions.cast
import net.auoeke.extensions.find
import net.auoeke.huntinghamhills.plugin.transformer.MethodTransformer
import net.auoeke.huntinghamhills.plugin.transformer.TransformerPlugin
import net.auoeke.limitless.enchantment.EnchantingBlocks
import net.auoeke.limitless.enchantment.EnchantmentUtil
import net.fabricmc.loader.api.FabricLoader
import net.minecraft.enchantment.Enchantment
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import org.objectweb.asm.Opcodes
import org.objectweb.asm.tree.*
import org.spongepowered.asm.mixin.transformer.LimitlessCoprocessor

class LimitlessTransformer : TransformerPlugin(), Opcodes {
    override fun onLoad(mixinPackage: String) {
        super.onLoad(mixinPackage)

        putField("creativeMode", 7477)
        putMethod("calculateRequiredExperienceLevel", 8227)
        putMethod("getMaxLevel", 8183)
    }

    override fun shouldApplyMixin(targetName: String, mixinName: String): Boolean = true.also {
        incompatibleMixins.forEach {
            if (FabricLoader.getInstance().isModLoaded(it.key) && it.value.matches(mixinName.substringAfter("net.auoeke.limitless.asm.mixin."))) {
                return false
            }
        }
    }

    @MethodTransformer(8229)
    private fun transformEnchantmentHelperGetPossibleEntries(method: MethodNode) {
        val instructions = method.instructions
        val iterator = instructions.iterator()

        iterator.find({it.opcode == Opcodes.INVOKEVIRTUAL && it.cast<MethodInsnNode>().name == this.method("getMaxLevel")}) {
            while (iterator.previous().type != AbstractInsnNode.LINE) {
                iterator.remove()
            }

            iterator.remove()

            for (instruction in iterator) {
                iterator.remove()

                if (instruction.type == AbstractInsnNode.IINC_INSN) {
                    break
                }
            }

            iterator.next()
            iterator.remove()
            instructions.insert(iterator.previous(), InstructionList()
                .iload(0)
                .aload(7)
                .aload(3)
                .invokestatic(
                    EnchantmentUtil.INTERNAL_NAME,
                    "getHighestSuitableLevel",
                    methodDescriptor('V', 'I', Enchantment::class, List::class),
                    false
                )
            )
        }
    }

    @MethodTransformer(17411)
    private fun transformEnchantmentScreenHandler(method: MethodNode) {
        val iterator = method.instructions.iterator()

        iterator.find({it.opcode == Opcodes.ICONST_0}) {instruction: InsnNode ->
            instruction.next.cast<VarInsnNode>().opcode = Opcodes.FSTORE

            iterator.set(MethodInsnNode(
                Opcodes.INVOKESTATIC,
                EnchantingBlocks.INTERNAL_NAME,
                "countEnchantingPower",
                methodDescriptor('F', World::class, BlockPos::class),
                false
            ))

            iterator.previous()
            iterator.add(VarInsnNode(Opcodes.ALOAD, 2))
            iterator.add(VarInsnNode(Opcodes.ALOAD, 3))
        }

        iterator.next()
        iterator.next()

        var gotoCount = 0

        while (gotoCount != 3) {
            if (iterator.next().opcode == Opcodes.GOTO) {
                ++gotoCount
            }

            iterator.remove()
        }

        iterator.find({instruction -> instruction.opcode == Opcodes.ILOAD && instruction.cast<VarInsnNode>().`var` == 4}) {instruction: VarInsnNode ->
            instruction.opcode = Opcodes.FLOAD
        }

        iterator.find({it.opcode == Opcodes.INVOKESTATIC}) {instruction: MethodInsnNode ->
            instruction.owner = EnchantingBlocks.INTERNAL_NAME
            instruction.name = "calculateRequiredExperienceLevel"
            instruction.desc = instruction.desc.replaceFirst("II".toRegex(), "IF")
            instruction.itf = false
        }
    }

    @MethodTransformer(8230)
    private fun transformEnchantmentHelperGenerateEnchantments(method: MethodNode) {
        val iterator = method.instructions.iterator()

        iterator.find({it.type == AbstractInsnNode.INT_INSN && (it as IntInsnNode).operand == 50}) {
            iterator.remove()
            iterator.add(VarInsnNode(Opcodes.ILOAD, 2))
            iterator.add(InsnNode(Opcodes.ICONST_2))
            iterator.add(InsnNode(Opcodes.IDIV))
            iterator.add(IntInsnNode(Opcodes.BIPUSH, 20))
            iterator.add(InsnNode(Opcodes.IADD))
        }
    }

    @MethodTransformer(7246, type = 1648)
    private fun transformEnchantBookFactoryCreate(method: MethodNode) {
        method.instructions.forEach {
            if (it.opcode == Opcodes.INVOKEVIRTUAL && (it as MethodInsnNode).name == getMaxLevel) {
                it.name = limitless_getOriginalMaxLevel
            }
        }
    }

    companion object {
        const val limitless_getOriginalMaxLevel = "limitless_getOriginalMaxLevel"

        val getMaxLevel: String get() = method(8183)
        val incompatibleMixins: Map<String, Regex> = mapOf(
            "taxfreelevels" to "normalization\\..*",
            "levelz" to "normalization\\.AnvilScreenHandlerMixin"
        ).mapValues {it.value.toRegex()}

        init {
            LimitlessCoprocessor.init()
        }
    }
}
